include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_differentiation_ad_ad_helpers_h
+#define dealii_differentiation_ad_ad_helpers_h
+
+#include <deal.II/base/config.h>
+
+#if defined(DEAL_II_WITH_ADOLC) || defined(DEAL_II_TRILINOS_WITH_SACADO)
+
+#  include <deal.II/base/numbers.h>
+
+#  include <deal.II/differentiation/ad/ad_drivers.h>
+#  include <deal.II/differentiation/ad/ad_number_traits.h>
+#  include <deal.II/differentiation/ad/adolc_number_types.h>
+#  include <deal.II/differentiation/ad/adolc_product_types.h>
+#  include <deal.II/differentiation/ad/sacado_number_types.h>
+#  include <deal.II/differentiation/ad/sacado_product_types.h>
+
+#  include <algorithm>
+#  include <iostream>
+#  include <iterator>
+#  include <numeric>
+#  include <set>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Differentiation
+{
+  namespace AD
+  {
+    /**
+     * A base helper class that facilitates the evaluation of the derivative(s)
+     * of a number of user-defined dependent variables $\mathbf{f}(\mathbf{X})$
+     * with respect to a set of independent variables $\mathbf{X}$, that is
+     * $\dfrac{d^{i} \mathbf{f}(\mathbf{X})}{d^{i} \mathbf{X}}$.
+     *
+     * In addition to the dimension @p dim, this class is templated on the
+     * floating point type @p scalar_type of the number that we'd like to
+     * differentiate, as well as an enumeration indicating the @p ADNumberTypeCode .
+     * The @p ADNumberTypeCode dictates which auto-differentiation library is
+     * to be used, and what the nature of the underlying auto-differentiable
+     * number is. Refer to the @ref auto_symb_diff module for more details in
+     * this regard.
+     *
+     * For all of the classes derived from this base class, there are two
+     * possible ways that the code in which they are used can be structured.
+     * The one approach is effectively a subset of the other, and which might
+     * be necessary to use depends on the nature of the chosen
+     * auto-differentiable number.
+     *
+     * When "tapeless" numbers are employed, the most simple code structure
+     * would be of the following form:
+     *
+     * @code
+     *   // Initialize AD helper
+     *   ADHelperType<dim,tapeless_AD_typecode> ad_helper (...);
+     *
+     *   // Register independent variables
+     *   ad_helper.register_independent_variable(...);
+     *
+     *   // Extract the sensitive equivalent of the independent variables
+     *   // The operations conducted with these numbers will be tracked.
+     *   const auto ad_independent_variables
+     *     = ad_helper.get_sensitive_variables(...);
```

Comment:
how does this work? how would the `ad_helper` know which of the independent variables are sensitive and which are not?

also, I wonder whether there is a way to define the term "sensitive". Maybe a glossary entry, and a link to it from the first place here (and maybe elsewhere) where you use it?

Reply:
By definition, all of the independent variables will be made "sensitive". Internally, we create a vector of AD numbers that are _marked_ as being independent variables. The values sent into the `register_independent_variable()` function above are used to initialise these AD counterparts. I've made a small note to this effect in the code examples.